cmd/internal/planetscale_edge_database.go

```diff
@@ -285,49 +279,92 @@ func (p PlanetScaleEdgeDatabase) sync(ctx context.Context, tc *psdbconnect.Table
+			switch event.Type {
+			case binlogdata.VEventType_VGTID:
+				vgtid := event.GetVgtid().ShardGtids[0]
+				if vgtid != nil {
+					// Update cursor to new VGTID
+					tc = &psdbconnect.TableCursor{
+						Shard:    tc.Shard,
+						Keyspace: tc.Keyspace,
+						Position: vgtid.Gtid,
+					}
+				}
+			case binlogdata.VEventType_LASTPK:
+				if event.LastPKEvent.TableLastPK != nil {
+					// Only update last PK because we're in a COPY phase
+					tc = &psdbconnect.TableCursor{
+						Shard:       tc.Shard,
+						Keyspace:    tc.Keyspace,
+						LastKnownPk: event.LastPKEvent.TableLastPK.Lastpk,
+						Position:    tc.Position,
+					}
+					p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sLASTPK event found, setting last PK to %+v", preamble, tc))
+				}
+			case binlogdata.VEventType_FIELD:
+				// Save fields for processing
+				fields = event.FieldEvent.Fields
+				p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sFIELD event found, setting fields to %+v", preamble, fields))
+			case binlogdata.VEventType_ROW:
+				// Collect rows for processing
+				for _, change := range event.RowEvent.RowChanges {
+					if change.After != nil {
+						rows = append(rows, &query.QueryResult{
+							Fields: fields,
+							Rows:   []*query.Row{change.After},
+						})
+					}
+				}
+			case binlogdata.VEventType_COPY_COMPLETED:
+				p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sCOPY_COMPLETED event found, copy phase finished", preamble))
+				copyCompletedSeen = true
+			}
 		}
 
-		// Heartbeats and other non-DML queries can create binlog events with the same VGTID, but no rows.
-		// These no-row results can have the same VGTID as a subsequent result with rows.
-		if len(res.Result) > 0 {
-			p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sFound %+v results", preamble, len(res.Result)))
-			// Watch for VGTID change as soon as we encounter records from some VGTID that is equal to, or after the stop position we're looking for.
-			// We watch for a VGTID that is equal to or after (not just equal to) the stop position, because by the time the first sync for records occurs,
-			// the current VGTID may have already advanced past the stop position.
-			watchForVgGtidChange = watchForVgGtidChange || positionAtLeast(tc.Position, stopPosition)
-			for _, result := range res.Result {
+		if waitForCopyCompleted && copyCompletedSeen {
+			p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sCopy phase completed. Waiting for next VGTID after stop position.", preamble))
+			watchForVgGtidChange = true
+		}
+		if !waitForCopyCompleted && positionEqual(tc.Position, stopPosition) {
+			p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sStop position [%+v] found. Waiting for next VGTID after stop position.", preamble, stopPosition))
+			watchForVgGtidChange = true
+		}
+
+		// Exit sync and flush records once the VGTID position is past the desired stop position, and we're no longer waiting for COPY phase to complete
+		if watchForVgGtidChange && positionAfter(tc.Position, stopPosition) {
```

Comment:
Once we know we can stop looking for rows, and the current position has advanced past the stop position, we can exit syncing.

Reply:
based on what this `watchForVgGtidChange` does, maybe a better variable name would be `watchForStopPosition`?